drivers/trone/TROne.cpp

```diff
@@ -0,0 +1,183 @@
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <string.h>
+#include "DriverFramework.hpp"
+#include "TROne.hpp"
+#ifdef __QURT
+#include "dev_fs_lib_i2c.h"
+#endif
+
+using namespace DriverFramework;
+
+int TROne::start()
+{
+	int result = devOpen(0);
+
+	/* Open the device path specified in the class initialization */
+	if (result < 0) {
+		DF_LOG_ERR("Unable to open the device path: %s", m_dev_path);
+		//goto exit;
+		return -1;
+	}
+
+	result = I2CDevObj::start();
+
+	if (result != 0) {
+		DF_LOG_ERR("error: could not start DevObj");
+		goto exit;
+	}
+
+	/* Configure the I2C bus parameters for the sensor. */
+	result = _setSlaveConfig(slave_addr,
+				 TRONE_BUS_FREQUENCY_IN_KHZ,
+				 TRONE_TRANSFER_TIMEOUT_IN_USECS);
+
+	if (result != 0) {
+		DF_LOG_ERR("I2C slave configuration failed");
+		goto exit;
+	}
+
+	result = DevObj::start();
+
+	if (result != 0) {
+		DF_LOG_ERR("error: could not start DevObj");
+		goto exit;
+	}
+
+	// probe to check if device is reachable
+	result = probe();
+
+exit:
+
+	if (result != 0) {
+		DF_LOG_ERR("error: Failed to start TROne");
+		DevObj::stop();
+		I2CDevObj::stop();
+		devClose();
+	}
+
+	return result;
+}
+
+int TROne::stop()
+{
+	int result = DevObj::stop();
+
+	if (result != 0) {
+		DF_LOG_ERR("DevObj stop failed");
+		return result;
+	}
+
+	usleep(100000);
+
+	result = devClose();
+
+	if (result != 0) {
+		DF_LOG_ERR("device close failed");
+		return result;
+	}
+
+	usleep(100000);
+	return 0;
+}
+
+int TROne::probe(void)
+{
+	int result = 0;
+	uint8_t who_am_i = 0;
+
+	result = _writeReg(TRONE_WHO_AM_I_REG, (uint8_t *)&who_am_i, sizeof(who_am_i));
+	if (result)
+	{
+		DF_LOG_ERR("Nonzero result when writing to TRONE_WHO_AM_I_REG 0x%x", TRONE_WHO_AM_I_REG);
+		return result;
+	}
+
+	result = simple_read((uint8_t *)&who_am_i, sizeof(who_am_i));
+
+	DF_LOG_INFO("WHO_AM_I bytes 0x%02x 0x%02x\n",
+		     (unsigned)who_am_i,
+		     TRONE_WHO_AM_I_REG_VAL);
```

Comment:
I think you want to make this `DF_LOG_DEBUG()` and instead add:

```
if ((unsigned)who_am_i != TRONE_WHO_AM_I_REG_VAL) {
    DF_LOG_ERR("Wrong WHO_AM_I: 0x%x instead of 0x%x", (unsigned)who_am_i, TRONE_WHO_AM_I_REG_VAL);
    return -1;
}
```